tests/integration/test_applications_send_logs.py

```diff
@@ -0,0 +1,127 @@
+#!/usr/bin/env python3
+# Copyright 2021 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import asyncio
+import logging
+import time
+from pathlib import Path
+
+import pytest
+import yaml
+from helpers import is_loki_up, loki_api_query, oci_image
+
+logger = logging.getLogger(__name__)
+
+METADATA = yaml.safe_load(Path("./metadata.yaml").read_text())
+resources = {"loki-image": METADATA["resources"]["loki-image"]["upstream-source"]}
+tester_resources = {
+    "workload-image": oci_image(
+        "./tests/integration/log-proxy-tester/metadata.yaml", "workload-image"
+    )
+}
+
+tester_apps = {
+    "loki-tester": "juju_application='loki-tester',level='debug'",
+    "log-proxy-tester-file": "juju_application='log-proxy-tester-file',filename=~'.+'",
+    "log-proxy-tester-syslog": "juju_appllication='log-proxy-tester-syslog',job=~'.*syslog'",
+}
+
+
+loki_app_name = "loki"
+tester_app_names = list(tester_apps.keys())
+app_names = [loki_app_name] + tester_app_names
+
+
+@pytest.mark.abort_on_fail
+async def test_loki_api_client_logs(
+    ops_test, loki_charm, loki_tester_charm, log_proxy_tester_charm
+):
+    """Test basic functionality of Loki push API relation interface."""
+    await asyncio.gather(
+        ops_test.model.deploy(
+            loki_charm,
+            resources=resources,
+            application_name=loki_app_name,
+        ),
+        ops_test.model.deploy(loki_tester_charm, application_name="loki-tester"),
+        ops_test.model.deploy(
+            log_proxy_tester_charm,
+            resources=tester_resources,
+            application_name="log-proxy-tester-file",
+        ),
+        ops_test.model.deploy(
+            log_proxy_tester_charm,
+            resources=tester_resources,
+            application_name="log-proxy-tester-syslog",
+            config={"syslog": "true", "file_forwarding": "false"},
+        ),
+    )
+    await ops_test.model.wait_for_idle(apps=app_names, status="active")
+
+    for t in tester_app_names:
+        await ops_test.model.add_relation(loki_app_name, t)
+    await ops_test.model.wait_for_idle(apps=app_names, status="active")
+
+    # This is silly, but we need to actually wait to get log data
+    time.sleep(30)
```

Comment:
Maybe we could `block_until` (with timeout) on a concrete log count ([http api](https://grafana.com/docs/loki/latest/api/#examples))?

Reply:
This patch literally uses the HTTP API.

No, this is not an option. The issue is that Loki endpoints are sometimes (frequently) sent _after_ `promtail` events fire, and I wanted to avoid modifying the library again so we could reconcile `LIBPATCH`. After more testing, emitting a `PromtailDigestError` in the case where a relation exists but there is no endpoint data and allowing the charm (or at least the tester charm) to go into BlockedStatus is unavailable.